### PR TITLE
[FLINK-16671][python] Support for defining scopes and variables on Python metric group

### PR DESCRIPTION
**Pull request by hequn8128**
## What is the purpose of the change

This pull request supports for defining scopes and variables on Python metric group, i.e., adding `get_metric_group()` on `FunctionContext`, adding `add_group`, `get_scope_components`, `get_all_variables` and `get_metric_identifier` in `MetricGroup`.

## Brief change log

  - Adds `get_metric_group()` on `FunctionContext`
  - Adds `add_group`, `get_scope_components`, `get_all_variables` and `get_metric_identifier` in `MetricGroup`
  - Adds `python.udf.metric.enabled` configuration to enable or disable metric for Python UDFs.

## Verifying this change

This change added tests and can be verified as follows:

  - Adds tests in `testUDFProtoConstructedProperlyForSingleUDF` to verify the metric group proto info.
  - Adds test_metric to verify operations in Python MetricGroup.
  - Adds test_metric_it_case to verify metric group info which is passed from Java to Python

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (PythonDocs, Note: more documentation will be added in [FLINK-16674](https://issues.apache.org/jira/browse/FLINK-16674))

### CODE REVIEW

File: flink-python/pyflink/metrics/metricbase.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+################################################################################
+import abc
+from enum import Enum
+
+
+class MetricGroup(abc.ABC):
+
+    def add_group(self, name: str, extra: str = None) -> 'MetricGroup':
+        """
+        if extra is not None, creates a new key-value MetricGroup pair. The key group
```
Comment (dianfu) [inline flink-python/pyflink/metrics/metricbase.py]: Should also add documentation about what will happen if extra is None.

File: flink-python/src/main/java/org/apache/flink/python/metric/FlinkMetricContainer.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import org.apache.flink.fnexecution.v1.FlinkFnApi;
+import org.apache.flink.metrics.MetricGroup;
+
+import java.util.Arrays;
+
+/**
+ * Helper class for forwarding metric group information from Java to Python and forward Python
+ * metrics to Java accumulators and metrics.
+ */
+public class FlinkMetricContainer {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/metric/FlinkMetricContainer.java]: Mark it as `@Internal`

File: flink-python/src/main/java/org/apache/flink/python/metric/FlinkMetricContainer.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+	public FlinkMetricContainer(MetricGroup metricGroup) {
+		this.baseMetricGroup = metricGroup;
+	}
+
+	public FlinkFnApi.MetricGroupInfo getBaseMetricGroupInfo() {
+		FlinkFnApi.MetricGroupInfo.Builder builder = FlinkFnApi.MetricGroupInfo.newBuilder();
+		// components
+		builder.addAllScopeComponents(
+			Arrays.asList(baseMetricGroup.getScopeComponents()));
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/metric/FlinkMetricContainer.java]: Replace with `Collections.singletonList`?

**Comment by hequn8128** [comment]
@dianfu Thanks a lot for your suggestions. The PR has been updated.

File: flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java [hunk 4]
```diff
@@ -55,7 +59,7 @@
 
 	private static final String MAIN_INPUT_ID = "input";
 
-	private final String taskName;
+	protected final String taskName;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java]: Unnecessary change

File: flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java [hunk 5]
```diff
@@ -116,17 +120,24 @@
 	 */
 	protected transient DataOutputViewStreamWrapper baosWrapper;
 
+	/**
+	 * The flinkMetricContainer will be set to null if metric is configured to be turned off.
+	 */
+	@Nullable protected FlinkMetricContainer flinkMetricContainer;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java]: Change to private?

File: flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java [hunk 6]
```diff
@@ -224,4 +235,15 @@ public JobBundleFactory createJobBundleFactory(Struct pipelineOptions) throws Ex
 	public abstract ExecutableStage createExecutableStage() throws Exception;
 
 	public abstract OutputReceiverFactory createOutputReceiverFactory();
+
+	/**
+	 * Gets the proto representation of the base MetricGroup used for all user-defined functions.
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java]: What about remove the `used for all user-defined functions` as metrics will be used not only for UDF?

File: flink-python/src/test/java/org/apache/flink/python/PythonOptionsTest.java [hunk 7]
```diff
@@ -95,4 +95,17 @@ public void testArrowBatchSize() {
 		final int actualArrowBatchSize = configuration.getInteger(PythonOptions.MAX_ARROW_BATCH_SIZE);
 		assertThat(actualArrowBatchSize, is(equalTo(expectedArrowBatchSize)));
 	}
+
+	@Test
+	public void testPthonUDFMetricEnabled() {
```
Comment (dianfu) [inline flink-python/src/test/java/org/apache/flink/python/PythonOptionsTest.java]: Rename to some other name like `testPythonMetric` or `testPythonMetricEnable`?

**Comment by hequn8128** [comment]
@dianfu Thanks a lot for your nice suggestions. The PR has been updated accordingly.

File: flink-python/pyflink/metrics/metricbase.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+    def add_group(self, name: str, extra: str = None) -> 'MetricGroup':
+        """
+        Creates a new MetricGroup and adds it to this groups sub-groups.
+
+        If extra is not None, creates a new key-value MetricGroup pair. The key group
+        is added to this groups sub-groups, while the value group is added to the key
+        group's sub-groups. This method returns the value group.
+
+        The only difference between calling this method and
```
Comment (dianfu) [inline flink-python/pyflink/metrics/metricbase.py]: The comments should be updated as get_all_variables doesn't exist any more

File: flink-python/src/main/java/org/apache/flink/python/PythonConfig.java [hunk 9]
```diff
@@ -103,6 +103,11 @@
 	@Nullable
 	private final String pythonExec;
 
+	/**
+	 * Whether metric is enabled for Python UDFs.
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/PythonConfig.java]: Remove `for Python UDFs.`?

File: flink-python/src/test/java/org/apache/flink/python/PythonOptionsTest.java [hunk 10]
```diff
@@ -95,4 +95,17 @@ public void testArrowBatchSize() {
 		final int actualArrowBatchSize = configuration.getInteger(PythonOptions.MAX_ARROW_BATCH_SIZE);
 		assertThat(actualArrowBatchSize, is(equalTo(expectedArrowBatchSize)));
 	}
+
+	@Test
+	public void testPthonMetricEnabled() {
```
Comment (dianfu) [inline flink-python/src/test/java/org/apache/flink/python/PythonOptionsTest.java]: typo: Pthon -> Python

**Comment by hequn8128** [comment]
The tests has been passed: https://travis-ci.org/github/hequn8128/flink/builds/667156612
